src/Dialect/Krnl/Krnl.td

```diff
@@ -180,26 +180,52 @@ def KrnlSeqExtractOp : Op<Krnl_Dialect, "seqextract", [MemRefsNormalizable,
     DeclareOpInterfaceMethods<AllocationOpInterface, ["buildDealloc", "buildClone"]>]> {
   let summary = "Krnl load from a seq";
   let description = [{
-    sequence is represented with memref<memref<>>.
-    This op loads a tensor for the sequence 'seq' at position 'index',
-    and return the tensor, which will be freed by Bufferization::Deallocation.
-    The element in the sequence will become null.
+    This op loads an element from the input sequence 'seq' at position 'index'.
+    The position value is guaranteed to be positive.
+    When the attribute 'copy' is 0: the extracted element is directly returned
+    When the attribute 'copy' is 1: the extracted element is copied and then return. 
+    The returned element is regarded as allocated by this Op because the 
+    element is created (copied) when inserted into the sequence.
+    Attribute 'copy' can be 0 when analysis or code generation guarantees 
+    that every element is extracted exactly once from the sequence.
   }];
   let arguments = (ins AnyMemRef:$seq,
-                       Index:$index);
+                       Index:$index,
+                       DefaultValuedAttr<SI64Attr, "0">:$copy);
 
   let results = (outs AnyType:$output);
 }
 
+def KrnlSeqInsertOp : Op<Krnl_Dialect, "seqinsert", [MemRefsNormalizable]> {
+  let summary = "Insert an element into a seq";
+  let description = [{
+    This op will create a new seq from the input seq by inserting the input
+    element at the desired position. The input element will be copied before 
+    inserted into the new seq, and the type will be casted to the type of seq 
+    element if needed.
+    The motivation to introduce this Op is to help bufferization::deallocation
+    The sequence is implemented as memref of memref in onnx-mlir.
+    When an element is stored into a sequence, its pointer is stored.
+    MLIR assumes values for Op, not pointer. As a result, 
+    The memref for the element will be freed after this Op if it is the last
+    use of the element.
+    To avoid dangling pointer, the element is copied before it is inserted 
+    into the sequence. The allocation for the copy will be revealed to graph
+    when the element is extracted from the sequence. Consequently, the 
+    deallocation pass can handle the space correctly.
+  }];
```

Comment:
Would it make sense to have a bool "copy" which an optimization can also determine to set to false if certain conditions are satisfied?

Reply:
No. Since the deallocation pass can not handle the pointer  stored in sequence, the element will be freed by deallocation if there is no other use. We may add traits for one shot optimization to transfer the space and avoid the copy in future.

Reply:
Deallocation will free the input element if there is no other usage. In future, we may use one-shot optimization pass to specify the transferring of content and avoid copying.

Reply:
So would it make sense to add the "copy" field now?

Reply:
No. Deallocation may be added before the element is extracted and used. Since the pointer for the element is stored to and loaded from the sequence. Deallocation pass can not see their u-d relationship.